test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from datetime import timedelta
+from typing import ClassVar
+from unittest import TestCase
+
+from torch.distributed import TCPStore
+
+from torch.distributed.elastic.rendezvous import RendezvousConnectionError, RendezvousParameters
+from torch.distributed.elastic.rendezvous.c10d_rendezvous_backend import (
+    C10dRendezvousBackend,
+    create_backend,
+)
+
+from rendezvous_backend_test import RendezvousBackendTestMixin
+
+
+class C10dRendezvousBackendTest(TestCase, RendezvousBackendTestMixin):
+    _store: ClassVar[TCPStore]
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls._store = TCPStore("localhost", 0, is_master=True)  # type: ignore[call-arg]
+
+    def setUp(self) -> None:
+        # Make sure we have a clean slate.
+        self._store.delete_key("torch.rendezvous.dummy_run_id")
+
+        self._backend = C10dRendezvousBackend(self._store, "dummy_run_id")
+
+    def _corrupt_state(self) -> None:
+        self._store.set("torch.rendezvous.dummy_run_id", "non_base64")
+
+
+class CreateBackendTest(TestCase):
+    def setUp(self) -> None:
+        self._params = RendezvousParameters(
+            backend="dummy_backend",
+            endpoint="localhost:29400",
+            run_id="dummy_run_id",
+            min_nodes=1,
+            max_nodes=1,
+            is_host="true",
+            store_type="tCp",
+            read_timeout="10",
+        )
+
+        self._expected_endpoint_host = "localhost"
+        self._expected_endpoint_port = 29400
+        self._expected_store_type = TCPStore
+        self._expected_read_timeout = timedelta(seconds=10)
+
+    def test_create_backend_returns_backend(self) -> None:
+        backend = create_backend(self._params)
+
+        self.assertIsInstance(backend.store, self._expected_store_type)
+
+        self.assertEqual(backend.name, "c10d-experimental")
+        self.assertEqual(backend.key, "torch.rendezvous." + self._params.run_id)
+
+        store = backend.store
+
+        self.assertEqual(store.host, self._expected_endpoint_host)  # type: ignore[attr-defined]
+        self.assertEqual(store.port, self._expected_endpoint_port)  # type: ignore[attr-defined]
+        self.assertEqual(store.timeout, self._expected_read_timeout)  # type: ignore[attr-defined]
+
+    def test_create_backend_returns_backend_if_is_host_is_false(self) -> None:
+        store = TCPStore(  # type: ignore[call-arg] # noqa: F841
+            self._expected_endpoint_host, self._expected_endpoint_port, is_master=True
+        )
+
+        self._params.config["is_host"] = "false"
```

Comment:
I'm a bit confused by this test, we end up creating the TCPStore twice here? Once on line 72 and then again in `test_create_backend_returns_backend` as part of create_backend? What exactly are we testing here that is not already tested by test_create_backend_returns_backend? Is it the fact that we can initialize two TCPStores on different ports?

Reply:
Also, how are we able to create two instances of TCPStore on the same port?

Reply:
> Also, how are we able to create two instances of TCPStore on the same port?

That is in fact what we are testing here. It simulates the case where more than one process on the same machine is part of the same job. Although quite an edge case in production, this can happen in test and development environments. Say for instance you instantiate a job with two workers on a single machine for testing purposes:

```
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 ...
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 ...
```

Because our endpoint is "localhost" both processes will attempt to instantiate the `TCPStore` with `is_master=True`. Obviously only one of them will be able to host the `TCPStore` and the other one will fail with something like "socket already in use". That second process will then re-attempt to instantiate the `TCPStore`, but this time it will explicitly pass `is_master=False`.

Note that this is the default behavior which would work in most cases. However if, for some reason, you want to override which exact process should be the store host, you can explicitly pass a `is_host` configuration option:

```
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 --args is_host=false ...
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 --args is_host=true ...
```

Now you know for sure that the second process will be the host of the store.